chore(deps): Upgrade backend infrastructure dependencies

fatih-acar opened the pull request:
## <!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added an automated workflow guide for upgrading backend infrastructure dependencies.

* **Chores**
  * Upgraded backend service images: Neo4j → 2025.10.1, Redis → 8.4.0, RabbitMQ → 4.2.1, PostgreSQL → 18.
  * Updated local demo and test environment configurations to use the newer images and updated volume path for Postgres.

* **Documentation**
  * Added changelog entry and updated demo docs to reflect upgraded dependencies.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

This PR upgrades backend infrastructure dependencies to their latest versions.

## Version Changes

| Component | Previous | New |
|-----------|----------|-----|
| Neo4j | 2025.03.0 | 2025.10.1 |
| Redis | 7.2.11 | 8.4.0 |
| RabbitMQ | 3.13.7 | 4.2.1 |
| PostgreSQL | 16-alpine | 18-alpine (pgautoupgrade) |

## Files Updated

- [x] `backend/tests/helpers/constants.py` - Neo4j image versions
- [x] `backend/tests/conftest.py` - Redis and RabbitMQ test container images
- [x] `docker-compose.yml` - Main compose file defaults
- [x] `development/docker-compose-deps.yml` - Development compose defaults
- [x] `development/docker-compose-deps-nats.yml` - NATS compose PostgreSQL
- [x] `tasks/shared.py` - Invoke task defaults
- [x] `python_testcontainers/infrahub_testcontainers/container.py` - Test container env vars
- [x] `python_testcontainers/infrahub_testcontainers/docker-compose.test.yml` - Test compose
- [x] `python_testcontainers/infrahub_testcontainers/docker-compose-cluster.test.yml` - Cluster test compose
- [x] `utilities/db_backup/__main__.py` - Backup utility Neo4j image
- [x] `.github/workflows/version-upgrade.yml` - CI workflow Neo4j image
- [x] `docs/docs/topics/local-demo-environment.mdx` - Documentation (fixed outdated versions)

## Testing

- [x] CI passes all tests
- [x] Local docker-compose starts successfully
- [x] Integration tests pass
- [x] No performance regression when running infrahub-private-tests (scale tests)
- [x] Migration and auto-upgrade of persistent containers pass (version-upgrade job succeeds)
- [x] Helm chart has been tested with upgraded dependencies

fatih-acar committed:
feat(claude): add upgrade backend dependencies command

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
chore(deps): upgrade backend infrastructure dependencies

- Neo4j: 2025.03.0 → 2025.10.1
- Redis: 7.2.11 → 8.4.0
- RabbitMQ: 3.13.7 → 4.2.1
- PostgreSQL: 16 → 18

Updated all relevant configuration files:
- Docker Compose files (main, dev, test, cluster)
- Test fixtures and constants
- GitHub workflows
- Documentation

fatih-acar committed:
fix(tests): update RabbitMQ test for 4.x compatibility

RabbitMQ 4.x reports queue types differently. Removed queue_type field
from Queue dataclass since it's not needed for the test assertions.

dgarros commented on .claude/commands/upgrade-backend-dependencies.md:
I think it would be best if we could have one PR per component by default,
it would be easier to identify regression / issues and troubleshoot.

Also if possible, I think it would be good to check the release note and identify if we need to change something in the code, and potentially open an issue to track that work if it's not mandatory to fix it right away

I understand the second point might be out of scope of this PR but I'd love to see us get there at some point

fatih-acar commented on .claude/commands/upgrade-backend-dependencies.md:
> I think it would be best if we could have one PR per component by default,
it would be easier to identify regression / issues and troubleshoot.

I will make the command ask for a specific component by default (instead of defaulting to ALL).
For this specific one, upgrading all components at once was useful to not waste time on multiple PRs (luckily it looks like there's no compat issue at all, only a small RabbitMQ change required)

> Also if possible, I think it would be good to check the release note and identify if we need to change something in the code, and potentially open an issue to track that work if it's not mandatory to fix it right away

I think the LLM can do most of this for us, I'll add a step in the command.

fatih-acar committed:
fix(ci): update the version-upgrade test job

Test for all dependencies upgrade.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
fix(docker-compose): postgres datadir changed in 18

Signed-off-by: Fatih Acar <fatih@opsmill.com>

fatih-acar committed:
fix(docker-compose): switch to pgautoupgrade image

This is a temporary change that will last during the next Infrahub
release so that users can seamlessly upgrade to the latest Postgres
version.
We will rollback to the original Postgres docker image during the N+1
Infrahub release.

Signed-off-by: Fatih Acar <fatih@opsmill.com>

ajtmccarty commented on .github/workflows/version-upgrade.yml:
leftover?

ogenstad commented on changelog/+upgrade-deps-20251206.changed.md:
This gets a bit weird as the versions aren't strictly tied to the code and anyone could run Infrahub with some undefined version of these backend services. It's correct in terms of what we default to throughout the repo and it's probably better to include a statement of the changes to the default services. But we could consider how or where this should be presented in the future.

ogenstad commented on .claude/commands/upgrade-backend-dependencies.md:
In general I'm not a fan of having the AI automatically create commits. In this case since it's so isolated to the updating of components it might be fine given the overall workflow of this. But outside of this when to commit changes to a branch and what to write is something that I want to be in control of.

fatih-acar commented on .github/workflows/version-upgrade.yml:
Indeed thanks!

fatih-acar commented on changelog/+upgrade-deps-20251206.changed.md:
I agree, this is because we are lucky enough to keep compatibility with older versions.
I believe we will need to add hard constraints to Infrahub at some point because of compat issues. Similar to what was done here: https://github.com/opsmill/infrahub/pull/6712

fatih-acar commented on .claude/commands/upgrade-backend-dependencies.md:
Shouldn't this be a global instruction to the repo? i.e. "never create commits" or "ask for confirmation before commit"

ogenstad commented on .claude/commands/upgrade-backend-dependencies.md:
That makes sense. Not sure if specific instructions in a command would override that though.
